### DISCUSSION TITLE
27.1 "GUI" library

### DISCUSSION THREAD

**Discussion by njr0** [google/comprehensive-rust#1266] [General]
It took me a little while to understand what we were supposed to do in this exercise. I think a couple of things confused me:

1. The button had a callback, which made me think we were actually going to implement something you could click and which would use the closure to print "You clicked the button!" But (unless I still haven't understood), there's none of that: it presumably just takes a callback because a real GUI would. But it threw me off.

2. It wasn't really clear to me at first what I could do with buffer because it wasn't quite similar enough to your example. In the end, I tried `write_all` on it which didn't work, and the compiler suggested `write_fmt`, from `std::format`, so I looked up the documentation for that and figured out that I could use `write!` (which I actually did) or `writeln!`, which seem to be what you want.

Maybe this is all good because not understanding and looking things up is all part of development anyway. And I did get there. But I feel like I got quite lucky by trying `write_all` (and `write`, in fact), and if I hadn't done that, I don't think I'd have known what to do.

**Comment by djmitche** [comment]
I agree regarding the callback. It'd be great to find a way to use `Box<dyn FnMut()>` in the example, but this isn't it. Making the example more complex to allow clicking on the button would add a lot of complexity and not work in the playground. So, we should probably remove that.

Perhaps the second point could be solved with a hint in a comment?
```rust
// HINT: use `write!(buffer, "a fmt string")` to write into a buffer.
```

**Comment by njr0** [comment]
Yes, I either remove the callback or put a comment right by it (probably in both places) pointing out that it's unused in this code.

Unless figuring out how to use `buffer` is a key part of the exercise (which it seems to me, it probably isn't) I think a hint like that would be helpful, yes.

**Reply by mgeisler** [reply]
Hey Nick! Yeah, I wish I had found a nice way to integrate a click on the button... but doesn't seem feasible so I agree we should remove it. Could you send a PR to clean it up?

**Comment by njr0** [comment]
Happy to take a crack at it, yes, but probably not today.

**Reply by mgeisler** [reply]
No rush! :slightly_smiling_face: Thanks for all the feedback! It is super valuable because you see the material through fresh eyes and you can thus point out many things I have not seen.

**Comment by njr0** [comment]
Likewise, thanks for the course. It's excellent (and I hope you realise the feedback is because I like it and want to improve it).

**Reply by mgeisler** [reply]
I'm very grateful for the feedback! Writing up detailed feedback takes more time than people think, so thank you for this!

The Rust community has been amazing in helping us improve the course! This is one of the things I tried to highlight in the recent [blog post about the course](https://security.googleblog.com/2023/09/scaling-rust-adoption-through-training.html).

**Comment by njr0** [comment]
So I was just looking at this and wondering what the best way to fix it is.

Presumably the callback isn't really right anyway: it's just using `println!`, but I'm guessing you'd really want to write that into the GUI/TUIsomewhere rather than printing it to the terminal or wherever.

So I think I'd be tempted to remove the callback completely and just add a note in the text that here we're only "drawing" a (textual) GUI and that in reality you'd pass a callback to the button. And I'm guessing that callback would actually need to be called with some kind of reference to the window or a widget it was going to write into and cause a refresh?

I'll do something brutal and you can see what you think.

**Reply by mgeisler** [reply]
Thanks for #1293, it looks good!

**Comment by njr0** [comment]
Done. See what you think.
